src/particles.zig

```diff
@@ -0,0 +1,413 @@
+const std = @import("std");
+const main = @import("main");
+const chunk_meshing = @import("renderer/chunk_meshing.zig");
+const graphics = @import("graphics.zig");
+const SSBO = graphics.SSBO;
+const TextureArray = graphics.TextureArray;
+const Shader = graphics.Shader;
+const Image = graphics.Image;
+const game = @import("game.zig");
+const ZonElement = @import("zon.zig").ZonElement;
+const c = graphics.c;
+const random = @import("random.zig");
+const vec = @import("vec.zig");
+const Mat4f = vec.Mat4f;
+const Vec3d = vec.Vec3d;
+const Vec4d = vec.Vec4d;
+const Vec3f = vec.Vec3f;
+const Vec4f = vec.Vec4f;
+const Vec3i = vec.Vec3i;
+
+var seed: u64 = undefined;
+
+var arena = main.heap.NeverFailingArenaAllocator.init(main.globalAllocator);
+const arenaAllocator = arena.allocator();
+
+pub const ParticleManager = struct {
+	var particleTypesSSBO: SSBO = undefined;
+	var types: main.List(ParticleType) = undefined;
+	var textures: main.List(Image) = undefined;
+	var emissionTextures: main.List(Image) = undefined;
+
+	var textureArray: TextureArray = undefined;
+	var emissionTextureArray: TextureArray = undefined;
+
+	const ParticleIndex = u16;
+	var particleTypeHashmap: std.StringHashMapUnmanaged(ParticleIndex) = undefined;
+
+	pub fn init() void {
+		types = .init(arenaAllocator);
+		textures = .init(arenaAllocator);
+		emissionTextures = .init(arenaAllocator);
+		textureArray = .init();
+		emissionTextureArray = .init();
+		ParticleSystem.init();
+	}
+
+	pub fn deinit() void {
+		types.deinit();
+		textures.deinit();
+		emissionTextures.deinit();
+		textureArray.deinit();
+		emissionTextureArray.deinit();
+		particleTypeHashmap.deinit(arenaAllocator.allocator);
+		ParticleSystem.deinit();
+		arena.deinit();
+	}
+
+	pub fn register(assetsFolder: []const u8, id: []const u8, zon: ZonElement) void {
+		const animationFrames = zon.get(u32, "animationFrames", 1);
+		const textureId = zon.get([]const u8, "texture", "cubyz:spark");
```

Comment:
I don't like this default, I think it should just fail and report that ID must be specified. I hate silent failures. Use null as default and use orelse to log and early return.